packages/api/amplify_api/lib/src/decorators/authorize_websocket_message.dart

```diff
@@ -0,0 +1,68 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:http/http.dart' as http;
+import 'package:meta/meta.dart';
+
+import '../graphql/ws/websocket_message.dart';
+import 'authorize_http_request.dart';
+
+/// Takes input websocket message (connection or subscription establisher) and
+/// adds authorization headers from auth repo.
+@internal
+Future<WebSocketMessage> authorizeWebSocketMessage(
+  WebSocketMessage inputMessage,
+  AWSApiConfig config,
+  AmplifyAuthProviderRepository authRepo,
+) async {
+  final body = inputMessage.payload?.toJson()['data'];
+  if (inputMessage is WebSocketConnectionInitMessage) {
+    inputMessage.authorizationHeaders =
+        await _generateAuthorizationHeaders(config, authRepo: authRepo);
+  } else if (body is String) {
+    inputMessage.payload?.authorizationHeaders =
+        await _generateAuthorizationHeaders(config,
+            authRepo: authRepo, body: body);
+  }
+  return Future.value(inputMessage);
```

Comment:
This function took me so long to grok, I would suggest reworking it so that the significance of each branch is more clear. Grabbing body on the first line and then only using it in the `else` branch makes no sense to me. It's also not clear what the significance of `body is String` is or why it's okay to return in some cases without adding authorization headers or why for one branch the authorization headers are on the message itself but then the other branch they're in the payload.

Creating an enum for the message types would be better than `is` checks, IMO, and would make this a lot more readable. Also linking to the docs would help give context to why authorization headers land in two different spots - or alternatively, put the authorization headers on the message itself for all types. 

I think finding a way to not modify input parameters directly would be good too, since 1) there's no reason to return the message if you're just modifying it in the function and 2) it would help avoid the confusion of where to put the authorization headers if this is a pure function.